### PR TITLE
Allow base64 encoded cookie content 

### PR DESCRIPTION
**Pull request by layoutd**
### Changes proposed in this Pull Request:

This PR adds a configuration parameter that allows the user to specify that cookie content should be base64 encoded. The objective is to reduce the chance that the cookie value will contain a value that triggers web application firewall rules (like `|||id=` or `ls …`).

_Notes:_
- The process of encoding values also removes any padding `=` at the end of the encoded value, leaving a clean alphanumeric value. (They're added again during decoding)
- When determining whether to decode from base64, we check for fully alphanumeric values. Any other value (`=`, `%`, default delimiter `|||`) indicates that the value is probably not base64-encoded.
- This also permits the configuration to be enabled or disabled without losing existing data:
  - Enabling will still read existing data as non-base64 until it's been saved again, encoded.
  - When enabled, it will try to determine cookie content as encoded or not and treat it accordingly.
  - Disabling it after having it enabled will detect encoded values and decode them.
  - When disabled, values that aren't encoded (have characters invalid for base64)
- This does means that some existing values won't be automatically encoded when the setting is enabled. Specifically, information about the first visit may be saved.

### How to test the changes in this Pull Request:
1. Install the repo on a web server
2. Visit the demo page (`/public/`) and check that everything works as expected: attribution is captured and persists after refresh.
3. View the `sbjs_*` cookies and confirm that they have the original format, URL encoded (`xxx=yyy|||aaa=bbb`, etc.):
    image
4. Enable base64 encoding in the demo:
    - `/src/html/index.sim:126` → `base64: true,`
    - `npm run build`
 5. Refresh the demo page ('/public/`) and confirm that all attribution remains available.
3. View the `sbjs_*` cookies and confirm that _some_ have been base64 encoded (`sbjs_current`, `sbjs_current_add`, `sbjs_session`, and `sbjs_udata`) and persists after refresh:
    image
3. Clear cookies and reload the demo page, and confirm that _all_ cookies are base64 encoded.
4. Visit one of the UTM test links at the bottom of the demo page and confirm that data is captured correctly and stored base64-encoded.
5. Disable base64 encoding the in the demo (see step 4, delete the `base64` config parameter and build and build).
6. Reload and confirm that all data persists.
7. Check the cookies and see that the updated cookies (`sbjs_current`, `sbjs_current_add`, `sbjs_session`, and `sbjs_udata`) are back to the classic URL-encoded format.
8. Clear cookies, refresh, and confirm data is captured, and all cookies are in the classic URL-encoded format.

### CODE REVIEW

**Review by rawdreeg — APPROVED** [review]
Really neat! Thanks @layoutd 

But I'd still let someone with better JS intuitions than me have a look :-)

File: README.md [hunk 1]
```diff
@@ -118,6 +119,10 @@ sbjs.init({
   // Set custom session length in minutes
   // 30 minutes is default
   session_length: 30,
+  
+  // Enable base64 encoding for cookie data
+  // disabled by default
+  params.base64 = true,
```
Comment (ecgan) [inline README.md]: Copy and paste error?

```suggestion
  base64: true,
```

File: src/js/helpers/cookies.js [hunk 2]
```diff
@@ -53,7 +64,14 @@ module.exports = {
       var c = ca[i];
       while (c.charAt(0) === ' ') { c = c.substring(1, c.length); }
       if (c.indexOf(nameEQ) === 0) {
-        return this.decodeData(c.substring(nameEQ.length, c.length));
+        var cookie_content = c.substring(nameEQ.length, c.length);
+        // Decode the content if it is only base64-valid characters
+        // Non-base64 characters indicate that the cookie is in the old format
+        if (/^[A-Za-z0-9+/]+$/.test(cookie_content)) {
+            // Base64 decoding after adding equal sign padding if necessary
+            cookie_content = atob(cookie_content.padEnd(cookie_content.length + (4 - cookie_content.length % 4) % 4, '='));
```
Comment (ecgan) [inline src/js/helpers/cookies.js]: We should add a try-catch block around `atob`, just in case.

Reference: https://developer.mozilla.org/en-US/docs/Web/API/atob#exceptions

Comment (ecgan) [reply src/js/helpers/cookies.js]: Also, is this formula correct?

```
cookie_content.padEnd(cookie_content.length + (4 - cookie_content.length % 4) % 4, '=')
```

From the formula, it might end up having 3 equal signs at the back `===`. From https://en.wikipedia.org/wiki/Base64, there should only be 0, 1 or 2 equal signs?

**Review by ecgan — COMMENTED** [review]
> * The process of encoding values also removes any padding `=` at the end of the encoded value, leaving a clean alphanumeric value. (They're added again during decoding)

❓ Why do we want to remove the padding `=` at the end of the encoded value? If we leave them intact, would it cause issues? 🤔 

I tested based on the test instructions. One of the steps is not working as expected. See below.

> 6\. View the `sbjs_*` cookies and confirm that _some_ have been base64 encoded (`sbjs_current`, `sbjs_current_add`, `sbjs_session`, and `sbjs_udata`) and persists after refresh:

❓ Why _some_ are automatically converted to become base64 encoded, while the others are not base64 encoded? I think this is because of the way the code currently works?

❓ In my testing, the `sbjs_current_add` is not base64 encoded? See screenshot below:

image

Other than the above step 6, everything else works as expected.

Leaving 3 comments below.

File: README.md (hunk 1 above)
Comment (layoutd) [reply README.md]: Nice catch, thanks

File: src/js/helpers/cookies.js (hunk 2 above)
Comment (layoutd) [reply src/js/helpers/cookies.js]: Added the try-catch block in https://github.com/woocommerce/sourcebuster-js/pull/6/commits/c938536f5e1e374ce3fd512876f5160bd9eb34b2

Comment (layoutd) [reply src/js/helpers/cookies.js]: > Also, is this formula correct?

While the formula _could_ technically add 3 `=` signs, all `btoa()` results have lengths that are multiples of 4, with 0, 1, or 2 `=` signs at the end. The formula ensures the final length is a multiple of 4 by checking how many characters are needed/missing (which results in adding only the missing, original 0, 1 or 2 `=` signs).

**Comment by layoutd** [comment]
Thanks as always, @ecgan!
> ❓ Why do we want to remove the padding `=` at the end of the encoded value? If we leave them intact, would it cause issues? 🤔

Just trying to be super cautious, I worry that some WAFs could detect [Base64-encoded strings as XSS or something](https://www.we45.com/post/how-to-prevent-xss-why-base64-is-not-enough). A simple alphanumeric string can't be misconstrued, right? RIGHT?

> ❓ Why _some_ are automatically converted to become base64 encoded, while the others are not base64 encoded? I think this is because of the way the code currently works?

Some cookies are only set on first visit and then not changed - `sbjs_first*` and `sbjs_migrations` and `sbjs_promo`, for example. The rest are updated on every page load, so they're read and/or overwritten with Base64 encoding after enabling the feature.
 
> ❓ In my testing, the `sbjs_current_add` is not base64 encoded? See screenshot below:

There are several conditions for `sbjs_current_add` to be updated: https://github.com/woocommerce/sourcebuster-js/blob/41c1391e7ed454996491b9ca3e31dd9b01121235/src/js/init.js#L31-L62

Basically, UTM params, organic source (Google), first pageload.

**Comment by layoutd** [comment]
> Why some are automatically converted to become base64 encoded, while the others are not base64 encoded? I think this is because of the way the code currently works?

I should add that there's also a migration functionality in SBJS, but that is aimed more at permanent changes, not changes in configuration (so, if in version 2.0 all values were _always_ Base64 encoded).

File: src/js/helpers/cookies.js (hunk 2 above)
Comment (ecgan) [reply src/js/helpers/cookies.js]: > While the formula _could_ technically add 3 `=` signs, all `btoa()` results have lengths that are multiples of 4, with 0, 1, or 2 `=` signs at the end. 

Yes, I agree with this part. Let's call this "part 1".

> The formula ensures the final length is a multiple of 4 by checking how many characters are needed/missing (which results in _adding only the missing_, original 0, 1 or 2 `=` signs).

Let's call this "part 2". I think I understand that you are seeing part 2 is linked with part 1. 

However, I'm seeing part 2 on its own without part 1. Or, in other words, I'm looking at just the `get` function and thinking how or where it might go wrong. For example, if someone or some codes change the `cookie_content` to become a single letter `"a"`, then our `padEnd` logic here will make the `cookie_content` become `"a==="`, which may cause issues. 

Anyway, I do think this is an unlikely case. I just want to voice it out (because it feels kinda weird looking at just part 2, without looking at part 1). We can choose to be more defensive and modify the formula, or we can leave it as is.

File: src/js/helpers/cookies.js [hunk 3]
```diff
@@ -68,8 +68,14 @@ module.exports = {
...
-            // Base64 decoding after adding equal sign padding if necessary
-            cookie_content = atob(cookie_content.padEnd(cookie_content.length + (4 - cookie_content.length % 4) % 4, '='));
+          try {
+            // Ensure the Base64 string has the correct padding (always multiples of 4)
+            cookie_content = cookie_content.padEnd(cookie_content.length + (4 - cookie_content.length % 4) % 4, '=');
+            // Decode the potentially padded Base64 string
+            cookie_content = atob(cookie_content);
+          } catch (error) {
+            // If the Base64 string is invalid, just keep the cookie content as is
+          }
```
Comment (ecgan) [inline src/js/helpers/cookies.js]: We are changing the `cookie_content` value, which I think might cause issues.

Imagine this scenario: 

1. At line 71, `cookie_content` is `abcde`. (Related to https://github.com/woocommerce/sourcebuster-js/pull/6#discussion_r1600371670, assuming this value is tampered by someone with dev tools or by some other codes.)
2. At line 73, after the `padEnd` logic, the `cookie_content` is set to `abcde===`.
3. At line 75, calling `atob(cookie_content)` will throw an error.
4. Notice that at this time, `cookie_content` is now `abcde===`, not the original value `abcde`. 
5. At line 80, we will call `this.decodeData(cookie_content)`, which is essentially `this.decodeData( 'abcde===' )`. Is this expected? Or should it be `this.decodeData( 'abcde' )`?
 
If this is not expected, we can fix it by assigning the padded value into a separate variable, so that we still have a reference to the original value.

**Review by ecgan — COMMENTED** [review]
Thanks @layoutd. Leaving one comment about your recent changes below, about changing the `cookie_content` value.

File: src/js/helpers/cookies.js (hunk 2 above)
Comment (ecgan) [reply src/js/helpers/cookies.js]: > Also, is this formula correct?

@layoutd , I just gave some thoughts about the formula, and I think of the following, what do you think?

```js
const padMe = (string) => {
  return string.padEnd(Math.ceil(string.length / 4) * 4, "=");
};
```

File: src/js/helpers/cookies.js (hunk 3 above)
Comment (layoutd) [reply src/js/helpers/cookies.js]: That's a good spot, top notch code review 🏅. 5 stars, would request again!
Changed in https://github.com/woocommerce/sourcebuster-js/pull/6/commits/25dcc9f48b42cffc2e8589fe2d1399f889f9d862

File: src/js/helpers/cookies.js (hunk 2 above)
Comment (layoutd) [reply src/js/helpers/cookies.js]: Yeah, that's 💯  cleaner, not sure what late night  Base64 investigation and trial-and-error resulted in my take on it 🤔. I've updated it in b507320b7dcaf96bfe35f0dc62d070471bbf76c1

**Comment by ecgan** [comment]
I wanted to comment about this but I forgot: 😅 

> > ❓ Why do we want to remove the padding `=` at the end of the encoded value? If we leave them intact, would it cause issues? 🤔
> 
> Just trying to be super cautious, I worry that some WAFs could detect [Base64-encoded strings as XSS or something](https://www.we45.com/post/how-to-prevent-xss-why-base64-is-not-enough). A simple alphanumeric string can't be misconstrued, right? RIGHT?

But a simple alphanumeric string like `bGlnaHQgd29y` is a valid base64-encoded string ([source](https://en.wikipedia.org/wiki/Base64)), so based on the concern you mentioned, the WAFs would be able to detect that. In other words, removing the `=` padding doesn't fully protect us from the WAFs scenario? If the WAFs block a string like `bGlnaHQgd29y` and do not block a string like `bGlnaHQgd29yaw` (with `==` padding at the end removed), then it makes issue reproduction and troubleshooting harder? If we want a better thorough protection, we may want to consider another approach? 🤔 

If we want to proceed with what we have now, I think it is probably fine (because the base64 setting is not the default, it requires "opt-in" via configuration parameter). Just want to throw out the questions for us to think about, and maybe we can improve on it later.

File: dist/sourcebuster.js [hunk 4]
```diff
@@ -169,18 +175,38 @@ module.exports = {
...
+        cookie_content = btoa(cookie_content).replace(/=+$/, '');
+    }
+    document.cookie = this.encodeData(name) + '=' + cookie_content + expires + basehost + '; path=/';
   },
 
   get: function(name) {
     var nameEQ = this.encodeData(name) + '=',
-        ca = document.cookie.split(';');
+        ca = document.cookie.split(';'),
+        padMe = function(string) {
```
Comment (ecgan) [inline dist/sourcebuster.js]: `padMe` was just a quick example name that I think of when I was testing the function, I didn't expect you to use the name as is. 😂

**Review by ecgan — APPROVED** [review]
Tested, works as expected, LGTM 👍

**Comment by layoutd** [comment]
> In other words, removing the `=` padding doesn't fully protect us from the WAFs scenario? If the WAFs block a string like `bGlnaHQgd29y` and do not block a string like `bGlnaHQgd29yaw`

This is a fair question, and my answer is, basically: I don't know. There are so many flavors of WAF and rulesets with minimal documentation that it's essentially a black box. Any change could or could not cause problems. For example, `|||id=`seems benign enough to not trigger a WAF rule either, but it does. Some rulesets look for `ls ` or `cd ` in cookie values. My basic logic here is just remove the "strange" characters like `=` at the end because, without that, the alphanumeric characters could be anything - API key, etc. 

In the end, idea here is to add an alternative that may provide a possible alternate solution for users whose WAFs have a problem with the default format of the data Sourcebuster stores in cookies.

File: dist/sourcebuster.js (hunk 4 above)
Comment (layoutd) [reply dist/sourcebuster.js]: I just trust the expert 🙄 ! Inlined in https://github.com/woocommerce/sourcebuster-js/pull/6/commits/f835c93439ad0045c68ebe8af84ff3d534ccf9ad

**Review by ecgan — APPROVED** [review]
LGTM 🙂 👍
